arraycontext/impl/pytato/utils.py

```diff
@@ -0,0 +1,97 @@
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+import numpy as np
+from typing import Any, Dict, Set, Tuple, Mapping
+from pytato.array import SizeParam, Placeholder as BasePlaceholder
+from pytato.array import ShapeType, TagsType, Array, DataWrapper
+from pytato.transform import CopyMapper
+from pytools import UniqueNameGenerator
+
+
+class Placeholder(BasePlaceholder):
+    """
+    An equality-comparable :class:`pytato.Placeholder`.
+    """
+    def __init__(self, name: str,
+                 shape: ShapeType,
+                 dtype: np.dtype,
+                 tags: TagsType = frozenset()
+                 ) -> None:
+        super().__init__(name, shape, dtype, tags)
+
+    def __eq__(self, other) -> bool:
+        return Array.__eq__(self, other)
+
+    def __hash__(self) -> int:
+        return Array.__hash__(self)
+
+
+class Normalizer(CopyMapper):
+    """
+    Helper mapper for :func:`normalize_pt_expr`. Every
+    :class:`pytato.DataWrapper` is replaced with a deterministic copy of
+    :class:`Placeholder`.
+    """
+    def __init__(self) -> None:
+        super().__init__()
+        self.bound_arguments: Dict[str, Any] = {}
+        self.vng = UniqueNameGenerator()
+        self.seen_inputs: Set[str] = set()
+
+    def map_data_wrapper(self, expr: DataWrapper) -> Array:
+        if expr.name is not None:
+            if expr.name in self.seen_inputs:
+                raise ValueError("Got multiple inputs with the name"
+                                 f"{expr.name} => Illegal.")
+            self.seen_inputs.add(expr.name)
+
+        name = self.vng("_actx_dw")
+        self.bound_arguments[name] = expr.data
+        return Placeholder(name=name,
+                           shape=tuple(self.rec(s) if isinstance(s, Array) else s
+                                       for s in expr.shape),
+                           dtype=expr.dtype,
+                           tags=expr.tags)
+
+    def map_size_param(self, expr: SizeParam) -> Array:
+        raise NotImplementedError
+
+    def map_placeholder(self, expr: BasePlaceholder) -> Array:
+        raise ValueError("Placeholders cannot appear in Normalizer.")
+
+
+def normalize_pt_expr(expr: Array) -> Tuple[Array,
+                                            Mapping[str, Any]]:
+    """
+    Returns ``(normalized_expr, bound_arguments)``.  *normalized_expr* is a
+    normalized form of *expr*, with all instances of
+    :class:`pytato.DataWrapper` replaced with instances of :class:`Placeholder`
+    in a deterministic manner. The data corresponding to the placeholders in
+    *normalized_expr* is recorded in the mapping *bound_arguments*.
+    """
+    normalize_mapper = Normalizer()
+    normalized_expr = normalize_mapper(expr)
+    return normalized_expr, normalize_mapper.bound_arguments
```

Comment:
Is this wrapper really necessary?

Reply:
Yep, see https://github.com/inducer/arraycontext/pull/56#discussion_r669565091. Otherwise we can (almost) never have 2 isomorphic DAGs being being equal.

Reply:
What I meant is not "we don't need this functionality", but rather "people can call the mapper directly".

Reply:
Adding those three lines to `PytatoPyOpenCLArrayContext.freeze` was looking a bit out-of-character. Made the routine an underscored one.